### PR TITLE
fix: implement a CacheStrategy to ensure compaction use cache correctly

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
Some changes revert https://github.com/GreptimeTeam/greptimedb/pull/5135/

## What's changed and what's your intention?

This PR implements a `CompactionStrategy` to replace `Option<CacheManagerRef>` so we can control how compaction uses caches.
- For queries, the strategy forwards all operations to the cache manager.
- The strategy only uses the write cache and the metadata cache for compactions.
- The `Disabled` strategy means disable all caches.

## PR Checklist
Please convert it to a draft if some of the following conditions are not met.

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.
- [ ] API changes are backward compatible.
- [ ] Schema or data changes are backward compatible.

### CODE REVIEW

File: src/mito2/src/cache.rs [hunk 1]
```diff
@@ -55,6 +55,193 @@ const FILE_TYPE: &str = "file";
...
+    /// Doesn't disable any cache.
+    Normal(CacheManagerRef),
+    /// Strategy for compaction.
+    /// Disables some caches that are not needed during compaction.
+    /// Enables the write cache so that the compaction can read files cached
+    /// in the write cache and write the compacted files back to the write cache.
+    Compaction(CacheManagerRef),
+    /// No cache.
+    Disabled,
+}
```
Comment (waynexia) [inline src/mito2/src/cache.rs:71]: naming strategy with target use scenarios seems misleading. Can we put how it actually works in the name? Like `EnableAll`, `MetaOnly`, `Disabled` (feel free to adjust these names)

Comment (evenyag) [reply src/mito2/src/cache.rs:71]: `EnableAll` is great. But the compaction needs a special cache strategy so I prefer to keep using `Compaction` as the strategy name.

**Comment by evenyag** [comment]
https://github.com/GreptimeTeam/greptimedb/actions/runs/12532626343/job/34951585899?pr=5254

I don't know why all runners already installed fuzz and gc-bin......
```
warning: tool `rust-analyzer` is already installed, remove it from `/home/runner/.cargo/bin`, then run `rustup update` to have rustup manage this tool.
warning: tool `rustfmt` is already installed, remove it from `/home/runner/.cargo/bin`, then run `rustup update` to have rustup manage this tool.
warning: tool `cargo-fmt` is already installed, remove it from `/home/runner/.cargo/bin`, then run `rustup update` to have rustup manage this tool.
    Updating crates.io index
 Downloading crates ...
  Downloaded cargo-fuzz v0.12.0
error: binary `cargo-fuzz` already exists in destination
Add --force to overwrite
    Updating crates.io index
 Downloading crates ...
  Downloaded cargo-gc-bin v0.1.4
error: binary `cargo-gc` already exists in destination
Add --force to overwrite
     Summary Failed to install cargo-fuzz, cargo-gc-bin (see error(s) above).
error: some crates failed to install
```

File: src/mito2/src/read/scan_region.rs [hunk 2]
```diff
@@ -582,7 +564,7 @@ impl ScanInput {
             predicate: None,
             memtables: Vec::new(),
             files: Vec::new(),
-            cache_manager: None,
+            cache_strategy: CacheStrategy::EnableAll(CacheManagerRef::default()),
```
Comment (v0y4g3r) [inline src/mito2/src/read/scan_region.rs:567]: Not sure if `CacheStrategy::EnableAll(CacheManagerRef::default())` makes any sense when `CacheManagerRef::default()` does not contain any cache?

**Review by v0y4g3r — COMMENTED** [review]
Rest LGTM

File: src/mito2/src/read/scan_region.rs (hunk 2 above)
Comment (evenyag) [reply src/mito2/src/read/scan_region.rs]: fixed in 58b83032f111191c5ec17abcff4c2aef2dc1ec29

**Review by waynexia — APPROVED** [review]

**Pull request merged**